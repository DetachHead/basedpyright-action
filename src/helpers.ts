import * as core from '@actions/core';
import * as httpClient from '@actions/http-client';
import * as tc from '@actions/tool-cache';
import * as path from 'path';
import SemVer from 'semver/classes/semver';
import stringArgv from 'string-argv';

import { NpmRegistryResponse } from './schema';

export function getNodeInfo() {
    return {
        version: process.version,
        execPath: process.execPath,
    };
}

export async function getArgs() {
    const pyrightVersion = await getPyrightVersion();
    const pyrightPath = await downloadPyright(pyrightVersion);

    const args = [pyrightPath];

    const workingDirectory = core.getInput('working-directory');

    const noComments = getBooleanInput('no-comments', false);
    if (!noComments) {
        args.push('--outputjson');
    }

    const pythonPlatform = core.getInput('python-platform');
    if (pythonPlatform) {
        args.push('--pythonplatform');
        args.push(pythonPlatform);
    }

    const pythonVersion = core.getInput('python-version');
    if (pythonVersion) {
        args.push('--pythonversion');
        args.push(pythonVersion);
    }

    const typeshedPath = core.getInput('typeshed-path');
    if (typeshedPath) {
        args.push('--typeshed-path');
        args.push(typeshedPath);
    }

    const venvPath = core.getInput('venv-path');
    if (venvPath) {
        args.push('--venv-path');
        args.push(venvPath);
    }

    const project = core.getInput('project');
    if (project) {
        args.push('--project');
        args.push(project);
    }

    const lib = getBooleanInput('lib', false);
    if (lib) {
        args.push('--lib');
    }

    const warnings = getBooleanInput('warnings', false);
    if (warnings) {
        args.push('--warnings');
    }

    const verifyTypes = core.getInput('verify-types');
    if (verifyTypes) {
        args.push('--verifytypes');
        args.push(verifyTypes);
    }

    const extraArgs = core.getInput('extra-args');
    if (extraArgs) {
        args.push(...stringArgv(extraArgs));
    }

    return {
        workingDirectory,
        noComments,
        pyrightVersion,
        args,
    };
}

function getBooleanInput(name: string, defaultValue: boolean): boolean {
    const input = core.getInput(name);
    if (!input) {
        return defaultValue;
    }
    return input.toUpperCase() === 'TRUE';
}

async function getPyrightVersion(): Promise<SemVer> {
    const versionSpec = core.getInput('version');
    if (versionSpec) {
        return new SemVer(versionSpec);
    }

    const client = new httpClient.HttpClient();
    const resp = await client.get('https://registry.npmjs.org/pyright/latest');
    const body = await resp.readBody();
    const obj = NpmRegistryResponse.parse(JSON.parse(body));
    return new SemVer(obj.version);
}

async function downloadPyright(version: SemVer): Promise<string> {
    // Note: this only works because the pyright package doesn't have any
    // dependencies. If this ever changes, we'll have to actually install it.
    const url = `https://registry.npmjs.org/pyright/-/pyright-${version.format()}.tgz`;
    const pyrightTarball = await tc.downloadTool(url);
    const pyright = await tc.extractTar(pyrightTarball);
    return path.join(pyright, 'package', 'index.js');
}
