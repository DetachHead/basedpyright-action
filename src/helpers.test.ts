import * as core from '@actions/core';
import * as httpClient from '@actions/http-client';
import * as tc from '@actions/tool-cache';

jest.mock('@actions/core');
const mockedCore = jest.mocked(core);
jest.mock('@actions/http-client');
const mockedHttpClient = jest.mocked(httpClient);
jest.mock('@actions/tool-cache');
const mockedTc = jest.mocked(tc);

import { getArgs } from './helpers';

beforeEach(() => {
    jest.clearAllMocks();
});

test('boolean input', () => {});
