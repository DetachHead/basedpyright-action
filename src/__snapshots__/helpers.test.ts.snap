// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getActionVersion: core.getInput 1`] = `[]`;

exports[`getActionVersion: tc.cacheDir 1`] = `[]`;

exports[`getActionVersion: tc.downloadTool 1`] = `[]`;

exports[`getActionVersion: tc.extractTar 1`] = `[]`;

exports[`getActionVersion: tc.find 1`] = `[]`;

exports[`getArgs bad version: core.getInput 1`] = `
[
  [
    "version",
  ],
]
`;

exports[`getArgs bad version: tc.cacheDir 1`] = `[]`;

exports[`getArgs bad version: tc.downloadTool 1`] = `[]`;

exports[`getArgs bad version: tc.extractTar 1`] = `[]`;

exports[`getArgs bad version: tc.find 1`] = `[]`;

exports[`getArgs valid version cached: core.getInput 1`] = `
[
  [
    "version",
  ],
  [
    "working-directory",
  ],
  [
    "no-comments",
  ],
  [
    "python-platform",
  ],
  [
    "python-version",
  ],
  [
    "typeshed-path",
  ],
  [
    "venv-path",
  ],
  [
    "project",
  ],
  [
    "lib",
  ],
  [
    "warnings",
  ],
  [
    "verify-types",
  ],
  [
    "ignore-external",
  ],
  [
    "extra-args",
  ],
]
`;

exports[`getArgs valid version cached: result 1`] = `
{
  "args": [
    "<TEMP_DIR>/rootDir/cached/pyright/package/index.js",
  ],
  "noComments": false,
  "pyrightVersion": "1.1.240",
  "workingDirectory": "",
}
`;

exports[`getArgs valid version cached: tc.cacheDir 1`] = `[]`;

exports[`getArgs valid version cached: tc.downloadTool 1`] = `[]`;

exports[`getArgs valid version cached: tc.extractTar 1`] = `[]`;

exports[`getArgs valid version cached: tc.find 1`] = `
[
  [
    "pyright",
    "1.1.240",
  ],
]
`;

exports[`getArgs valid version many options: core.getInput 1`] = `
[
  [
    "version",
  ],
  [
    "working-directory",
  ],
  [
    "no-comments",
  ],
  [
    "python-platform",
  ],
  [
    "python-version",
  ],
  [
    "typeshed-path",
  ],
  [
    "venv-path",
  ],
  [
    "project",
  ],
  [
    "lib",
  ],
  [
    "warnings",
  ],
  [
    "verify-types",
  ],
  [
    "ignore-external",
  ],
  [
    "extra-args",
  ],
]
`;

exports[`getArgs valid version many options: result 1`] = `
{
  "args": [
    "<TEMP_DIR>/rootDir/cached/pyright/package/index.js",
    "--pythonplatform",
    "Linux",
    "--pythonversion",
    "3.9",
    "--typeshed-path",
    "/path/to/typeshed",
    "--venv-path",
    "/path/to-venv",
    "--project",
    "/path/to/pyrightconfig.json",
    "--lib",
    "--warnings",
    "--foo",
    "--bar",
    "--baz",
  ],
  "noComments": false,
  "pyrightVersion": "1.1.240",
  "workingDirectory": "/path/to/project",
}
`;

exports[`getArgs valid version many options: tc.cacheDir 1`] = `
[
  [
    "<TEMP_DIR>/rootDir/pyright",
    "pyright",
    "1.1.240",
  ],
]
`;

exports[`getArgs valid version many options: tc.downloadTool 1`] = `
[
  [
    "https://registry.npmjs.org/pyright/-/pyright-1.1.240.tgz",
  ],
]
`;

exports[`getArgs valid version many options: tc.extractTar 1`] = `
[
  [
    "<TEMP_DIR>/rootDir/pyright.tar.gz",
  ],
]
`;

exports[`getArgs valid version many options: tc.find 1`] = `
[
  [
    "pyright",
    "1.1.240",
  ],
]
`;

exports[`getArgs valid version no comments: core.getInput 1`] = `
[
  [
    "version",
  ],
  [
    "working-directory",
  ],
  [
    "no-comments",
  ],
  [
    "python-platform",
  ],
  [
    "python-version",
  ],
  [
    "typeshed-path",
  ],
  [
    "venv-path",
  ],
  [
    "project",
  ],
  [
    "lib",
  ],
  [
    "warnings",
  ],
  [
    "verify-types",
  ],
  [
    "ignore-external",
  ],
  [
    "extra-args",
  ],
]
`;

exports[`getArgs valid version no comments: result 1`] = `
{
  "args": [
    "<TEMP_DIR>/rootDir/cached/pyright/package/index.js",
  ],
  "noComments": true,
  "pyrightVersion": "1.1.240",
  "workingDirectory": "",
}
`;

exports[`getArgs valid version no comments: tc.cacheDir 1`] = `
[
  [
    "<TEMP_DIR>/rootDir/pyright",
    "pyright",
    "1.1.240",
  ],
]
`;

exports[`getArgs valid version no comments: tc.downloadTool 1`] = `
[
  [
    "https://registry.npmjs.org/pyright/-/pyright-1.1.240.tgz",
  ],
]
`;

exports[`getArgs valid version no comments: tc.extractTar 1`] = `
[
  [
    "<TEMP_DIR>/rootDir/pyright.tar.gz",
  ],
]
`;

exports[`getArgs valid version no comments: tc.find 1`] = `
[
  [
    "pyright",
    "1.1.240",
  ],
]
`;

exports[`getArgs valid version verifytypes: core.getInput 1`] = `
[
  [
    "version",
  ],
  [
    "working-directory",
  ],
  [
    "no-comments",
  ],
  [
    "python-platform",
  ],
  [
    "python-version",
  ],
  [
    "typeshed-path",
  ],
  [
    "venv-path",
  ],
  [
    "project",
  ],
  [
    "lib",
  ],
  [
    "warnings",
  ],
  [
    "verify-types",
  ],
  [
    "ignore-external",
  ],
  [
    "extra-args",
  ],
]
`;

exports[`getArgs valid version verifytypes: result 1`] = `
{
  "args": [
    "<TEMP_DIR>/rootDir/cached/pyright/package/index.js",
    "--verifytypes",
    "some.package",
    "--ignoreexternal",
  ],
  "noComments": false,
  "pyrightVersion": "1.1.240",
  "workingDirectory": "",
}
`;

exports[`getArgs valid version verifytypes: tc.cacheDir 1`] = `
[
  [
    "<TEMP_DIR>/rootDir/pyright",
    "pyright",
    "1.1.240",
  ],
]
`;

exports[`getArgs valid version verifytypes: tc.downloadTool 1`] = `
[
  [
    "https://registry.npmjs.org/pyright/-/pyright-1.1.240.tgz",
  ],
]
`;

exports[`getArgs valid version verifytypes: tc.extractTar 1`] = `
[
  [
    "<TEMP_DIR>/rootDir/pyright.tar.gz",
  ],
]
`;

exports[`getArgs valid version verifytypes: tc.find 1`] = `
[
  [
    "pyright",
    "1.1.240",
  ],
]
`;

exports[`getArgs valid version version not found: core.getInput 1`] = `
[
  [
    "version",
  ],
]
`;

exports[`getArgs valid version version not found: tc.cacheDir 1`] = `[]`;

exports[`getArgs valid version version not found: tc.downloadTool 1`] = `[]`;

exports[`getArgs valid version version not found: tc.extractTar 1`] = `[]`;

exports[`getArgs valid version version not found: tc.find 1`] = `[]`;

exports[`getNodeInfo: core.getInput 1`] = `[]`;

exports[`getNodeInfo: tc.cacheDir 1`] = `[]`;

exports[`getNodeInfo: tc.downloadTool 1`] = `[]`;

exports[`getNodeInfo: tc.extractTar 1`] = `[]`;

exports[`getNodeInfo: tc.find 1`] = `[]`;
